**DataFrames Day 2**

Learning Outcomes: Prep for LO 1. Students will be able to select, run, and interpret univariate and bivariate statistical analyses and visualizations.

In [2]:
import pandas as pd
df = pd.read_csv('https://www.dropbox.com/s/puruptejplzhts5/insurance.csv?dl=1')
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

# **Filtering DataFrame By Column**

Return the age and charges columns

In [4]:
filtered_df = df[['age', 'charges', 'sex']] #need an extra set of brackets '[]'. Indicates we're inputting a list.
filtered_df.head().sort_values('charges', ascending=False)

age      charges     sex
3   33  21984.47061    male
0   19  16884.92400  female
2   28   4449.46200    male
4   32   3866.85520    male
1   18   1725.55230    male

## **Filtering by Row**

df[conditional] - sometimes we call that conditional a "mask" and it returns true and false values.

Useful instead of deleting

**Conditional Operations**

Equals : ==

Not equals : !=

Greater than: >

Less than: <

And: &

Or : |

is in a list: .isin(list)

is not in a list: ~, .isin(list) (see example)

If you use an & or an |, use () to group items together.

**Find all records where the individual is greater than 60**

In [5]:
df.age > 60 #Series of true and false values

0       False
1       False
2       False
3       False
4       False
        ...  
1333    False
1334    False
1335    False
1336    False
1337     True
Name: age, Length: 1338, dtype: bool

In [6]:
df[df.age > 60]

age     sex     bmi  children smoker     region      charges
11     62  female  26.290         0    yes  southeast  27808.72510
26     63  female  23.085         0     no  northeast  14451.83515
33     63    male  28.310         0     no  northwest  13770.09790
36     62  female  32.965         3     no  northwest  15612.19335
62     64    male  24.700         1     no  northwest  30166.61817
...   ...     ...     ...       ...    ...        ...          ...
1301   62    male  30.875         3    yes  northwest  46718.16325
1321   62    male  26.695         0    yes  northeast  28101.33305
1322   62    male  38.830         0     no  southeast  12981.34570
1325   61    male  33.535         0     no  northeast  13143.33665
1337   61  female  29.070         0    yes  northwest  29141.36030

[91 rows x 7 columns]

In [7]:
df[df['age'] > 60]


age     sex     bmi  children smoker     region      charges
11     62  female  26.290         0    yes  southeast  27808.72510
26     63  female  23.085         0     no  northeast  14451.83515
33     63    male  28.310         0     no  northwest  13770.09790
36     62  female  32.965         3     no  northwest  15612.19335
62     64    male  24.700         1     no  northwest  30166.61817
...   ...     ...     ...       ...    ...        ...          ...
1301   62    male  30.875         3    yes  northwest  46718.16325
1321   62    male  26.695         0    yes  northeast  28101.33305
1322   62    male  38.830         0     no  southeast  12981.34570
1325   61    male  33.535         0     no  northeast  13143.33665
1337   61  female  29.070         0    yes  northwest  29141.36030

[91 rows x 7 columns]

You must use the [] when referencing a column name that has a space or when using a variable to reference a column name.

If you want to save the filter for future work, you need to save it to a variable.

In [8]:
df_filtered = df[df.age > 60]
df_filtered.head()

age     sex     bmi  children smoker     region      charges
11   62  female  26.290         0    yes  southeast  27808.72510
26   63  female  23.085         0     no  northeast  14451.83515
33   63    male  28.310         0     no  northwest  13770.09790
36   62  female  32.965         3     no  northwest  15612.19335
62   64    male  24.700         1     no  northwest  30166.61817

**Return all individuals in the southwest region that have two or more children**

In [9]:
df[(df.region == 'southwest') & (df.children >=2)]

age     sex   bmi  children smoker     region    charges
29     31    male  36.3         2    yes  southwest  38711.000
32     19  female  28.6         5     no  southwest   4687.797
66     61  female  39.1         2     no  southwest  14235.072
72     53  female  28.1         3     no  southwest  11741.726
74     44    male  27.4         2     no  southwest   7726.854
...   ...     ...   ...       ...    ...        ...        ...
1288   20    male  39.4         2    yes  southwest  38344.566
1309   41    male  32.2         2     no  southwest   6875.961
1313   19  female  34.7         2    yes  southwest  36397.576
1329   52    male  38.6         2     no  southwest  10325.206
1332   52  female  44.7         3     no  southwest  11411.685

[109 rows x 7 columns]

**Return all individuals from the northwest region that are either 20 or younger or 60 and older. Sort by age descending.**

In [10]:
df[(df['region'] == 'northwest') & ((df.age <= 20) | (df.age >= 60))].sort_values(by=['age'], ascending = False)

age     sex     bmi  children smoker     region      charges
752   64    male  37.905         0     no  northwest  14210.53595
62    64    male  24.700         1     no  northwest  30166.61817
378   64  female  30.115         3     no  northwest  16455.70785
890   64  female  26.885         0    yes  northwest  29330.98315
402   64  female  32.965         0     no  northwest  14692.66935
..   ...     ...     ...       ...    ...        ...          ...
238   19    male  29.070         0    yes  northwest  17352.68030
773   19  female  28.880         0    yes  northwest  17748.50620
807   19  female  36.575         0     no  northwest   2136.88225
200   19  female  32.110         0     no  northwest   2130.67590
559   19    male  35.530         0     no  northwest   1646.42970

[69 rows x 7 columns]

**Return all individuals that have charges of 50000 dollars or greater and are from either the northwest, southwest, or southeast region.**

In [11]:
df[((df.region == 'northwest') | (df.region == "southwest") | (df.region == 'southeast')) & (df.charges >= 50000)]

age     sex     bmi  children smoker     region      charges
34     28    male  36.400         1    yes  southwest  51194.55914
543    54  female  47.410         0    yes  southeast  63770.42801
819    33  female  35.530         0    yes  northwest  55135.40209
1146   60    male  32.800         0    yes  southwest  52590.82939
1230   52    male  34.485         3    yes  northwest  60021.39897
1300   45    male  30.360         0    yes  southeast  62592.87309

In [12]:
regionList = ['northwest', 'southwest', 'southeast']
df[(df.region.isin(regionList)) & (df.charges >= 50000)]

age     sex     bmi  children smoker     region      charges
34     28    male  36.400         1    yes  southwest  51194.55914
543    54  female  47.410         0    yes  southeast  63770.42801
819    33  female  35.530         0    yes  northwest  55135.40209
1146   60    male  32.800         0    yes  southwest  52590.82939
1230   52    male  34.485         3    yes  northwest  60021.39897
1300   45    male  30.360         0    yes  southeast  62592.87309

**Return all individuals that have charges 50000 or greater and are NOT from either the northwest, southwest, or southeast region.**

In [13]:
df[(~df.region.isin(regionList)) & (df.charges >= 50000)]

age     sex     bmi  children smoker     region      charges
577   31  female  38.095         1    yes  northeast  58571.07448

# **Let's explore diamonds!**

In [1]:
import pandas as pd
df = pd.read_csv('https://www.dropbox.com/s/17mfhd6s4xc3cmk/Diamond%20Price%20Prediction.csv?dl=1')
df.head()

Carat      Cut Color Clarity  Depth  Table  Price  X(length)  Y(width)  \
0   0.23    Ideal     E     SI2   61.5   55.0    326       3.95      3.98   
1   0.21  Premium     E     SI1   59.8   61.0    326       3.89      3.84   
2   0.23     Good     E     VS1   56.9   65.0    327       4.05      4.07   
3   0.29  Premium     I     VS2   62.4   58.0    334       4.20      4.23   
4   0.31     Good     J     SI2   63.3   58.0    335       4.34      4.35   

   Z(Depth)  
0      2.43  
1      2.31  
2      2.31  
3      2.63  
4      2.75

[Here](https://www.kaggle.com/datasets/shivam2503/diamonds) is a key for what each column and value represent in the Diamond dataset.

Imagine you just started a job selling diamonds and the diamond dataset contains your current inventory. You want to prepare for potential customer questions by exploring your diamond dataset.

Below are a list of questions or scenerios you anticipate you will encounter.

**Q1. What are the characteristics of the highest priced diamond?**


In [2]:
df.sort_values(by= 'Price', ascending= False).head(1)

Carat      Cut Color Clarity  Depth  Table  Price  X(length)  Y(width)  \
27749   2.29  Premium     I     VS2   60.8   60.0  18823        8.5      8.47   

       Z(Depth)  
27749      5.16

**Q2. What are the characteristics of the lowest priced diamond?**



In [3]:
df.sort_values(by='Price', ascending=True).head(1)

Carat    Cut Color Clarity  Depth  Table  Price  X(length)  Y(width)  \
0   0.23  Ideal     E     SI2   61.5   55.0    326       3.95      3.98   

   Z(Depth)  
0      2.43

**Q3. You imagine some customers will be on a budget, but still want a nice cut diamond. Return all diamonds that have a cut of Very Good, Premium, or Ideal, have a carat of at least 0.6, and are less than 1000 dollars.**

In [5]:
good = ['Very Good', 'Premium', 'Ideal']
df[(df.Cut.isin(good)) & (df.Carat >= 0.6) & (df.Price < 1000)]


Carat      Cut Color Clarity  Depth  Table  Price  X(length)  Y(width)  \
32833   0.60  Premium     H      I1   60.1   60.0    806       5.51      5.44   
33664   0.60  Premium     H      I1   62.1   62.0    835       5.33      5.27   
36237   0.62  Premium     G      I1   58.9   60.0    933       5.65      5.54   
36571   0.72  Premium     J      I1   62.9   59.0    945       5.80      5.63   

       Z(Depth)  
32833      3.29  
33664      3.30  
36237      3.29  
36571      3.58

**Q4. You imagine some customers will want the very best diamond, regardless of price. Return all diamonds that have a carate of at least 1, and have the highest quality of cut, best color, *and* best clarity. You may need to review the data key above.**

In [6]:
df[(df.Carat >= 1) & (df.Cut == 'Ideal') & (df.Color == 'D') & (df.Clarity == 'IF')]

Carat    Cut Color Clarity  Depth  Table  Price  X(length)  Y(width)  \
25622   1.04  Ideal     D      IF   61.8   57.0  14494       6.49      6.52   
25718   1.04  Ideal     D      IF   61.8   57.0  14626       6.52      6.49   
26198   1.02  Ideal     D      IF   63.0   57.0  15575       6.39      6.35   
26311   1.06  Ideal     D      IF   61.2   57.0  15813       6.57      6.61   
26660   1.00  Ideal     D      IF   60.7   57.0  16469       6.44      6.48   
26965   1.07  Ideal     D      IF   60.9   54.0  17042       6.66      6.73   
27226   1.03  Ideal     D      IF   62.0   56.0  17590       6.55      6.44   

       Z(Depth)  
25622      4.02  
25718      4.02  
26198      4.01  
26311      4.03  
26660      3.92  
26965      4.08  
27226      4.03

**Q5.**A customer has already sent in an online inquiry for a specific diamond. The customer is coming into your store later today and you want to be prepared with some options.

**The customer wants to see diamonds that have a carat between 1 - 1.2, have an ideal cut, and are either D or E color. In addition, the customer wants to see diamonds with a carat between 1.6 and 2, regardess of the cut or color. The customer's price limit is 5,000.**

Sort by carat descending.


In [27]:
df[((df.Carat >= 1) & (df.Carat <= 1.2)) & (df.Cut == 'Ideal') & ((df.Color == "D") | (df.Color =="E"))].sort_values('Carat', ascending=False)

Carat    Cut Color Clarity  Depth  Table  Price  X(length)  Y(width)  \
22174    1.2  Ideal     E     VS2   61.3   56.0  10214       6.89      6.84   
13302    1.2  Ideal     E     SI2   61.1   55.0   5494       6.90      6.87   
18911    1.2  Ideal     D     SI1   62.7   55.0   7761       6.71      6.79   
18458    1.2  Ideal     D     SI1   61.8   58.0   7508       6.76      6.83   
15296    1.2  Ideal     D     SI2   60.2   57.0   6140       6.88      6.93   
...      ...    ...   ...     ...    ...    ...    ...        ...       ...   
11091    1.0  Ideal     D     SI1   62.5   55.0   4928       6.43      6.38   
11186    1.0  Ideal     D     SI2   61.5   56.0   4956       6.45      6.41   
14753    1.0  Ideal     E     SI1   61.7   55.0   5938       6.44      6.49   
14738    1.0  Ideal     E     SI1   62.9   56.0   5935       6.33      6.38   
3272     1.0  Ideal     E      I1   63.0   57.0   3360       6.35      6.32   

       Z(Depth)  
22174      4.21  
13302      4.21  
18911      4.23  
18458      4.20  
15296      4.16  
...         ...  
11091      4.00  
11186      3.95  
14753      3.99  
14738      4.00  
3272       3.99  

[679 rows x 10 columns]

# **Functions**

Convert 90 degrees fahrenheit to celsius. Don't use a function.

The formula to convert F to C is (fahrenheit temp - 32) * 0.5556

In [28]:
f=90
c = (f-32) * 0.5556

print(c)

32.2248


Create a function that converts a fahrenheit temperature to celsius. Pass the fahrenheit temperature as the parameter and return the celsius temperature.

In [29]:
def FtoC(f):
  c = (f-32) * 0.5556
  return c

Call the function

In [30]:
FtoC(90)

32.2248

# **You Try - Functions**

Create a function that will sort a column in a dataframe. The function should accept three parameters: a dataframe, column name, and either True or False. If True is passed, filter the column ascedending. If False is passed, filter the column descending.

In [34]:
def sort (df, colName, bool) :
    return df.sort_values(by=colName, ascending= bool)
    

df.sort(df, 'Price', True)

AttributeError: 'DataFrame' object has no attribute 'sort'

: 

# **Group By**

We may not have time to go over group bys, and you will learn this in your 402 class. But, we can make groups for values in a DataFrame column. Doing this we can then get aggregate information about another column.

In [5]:
#Group by smokers and non-smokers
#Get the count for each column
df.groupby('smoker').count()

age   sex   bmi  children  region  charges
smoker                                             
no      1064  1064  1064      1064    1064     1064
yes      274   274   274       274     274      274

In [10]:
#Group by smokers and non-smokers
#Get the average for the charges column
df.groupby('smoker').charges.mean()

smoker
no      8434.268298
yes    32050.231832
Name: charges, dtype: float64

# **Group By - You Try** 

Which region has the highest average number of children?